### ISSUE TITLE
Unable to Slice Out A Single Token From All Instances

### ISSUE THREAD

**Issue by KuanWeiBeCool** [shap/shap#2634]
Hi there, I would like to slice out the SHAP value for a single token from all instances for analysis, like what is shown in the documentation:
![image](https://user-images.githubusercontent.com/64152297/181776931-02ba0216-70b1-4d63-ba40-9ba3be373cf8.png)

However, I am unable to do the same thing for my SHAP values. My SHAP values has a shape of (4, None, 665), which means 4 instances, and each instance has different tokens thus the second is None, 665 labels. I tried the following ways to the all SHAP values for the token 'mix':

```
shap_values[:, 'mix', 365]
shap_values[:, :, 365][:, 'mix']
shap_values[0][:, 365]['mix']
```

The first two gives error `IndexError: too many indices for array`, the third gives error `IndexError: tuple index out of range`. Could anyone tell me the right way to do it? Thanks!

**Comment by remo-help** [comment]
Have you tried `shap.plots.bar(shap_values[:, :, 'mix'])`?

**Comment by KuanWeiBeCool** [comment]
Yes, it shows the following error message:
![image](https://user-images.githubusercontent.com/64152297/182189006-08baa21a-07d5-4bda-8d73-d6b35a04eb5d.png)
![image](https://user-images.githubusercontent.com/64152297/182189160-afdd6c1f-6123-4929-ba60-bcbd92146318.png)

**Comment by remo-help** [comment]
if you call `type(shap_values)` is the output a `shap.Explanation` object, or are you looking at an actual array of shap values?

**Comment by KuanWeiBeCool** [comment]
It's a shap.Explanation object
![image](https://user-images.githubusercontent.com/64152297/183093544-bf9b1bde-fe11-4a4a-b8eb-40e8ef57085e.png)

**Comment by remo-help** [comment]
Can you try the following and post the outputs:
```
shap_values.values.shape
shap_values.data.shape
shap_values.feature_names.shape
shap_values.base_values.shape
```

**Comment by KuanWeiBeCool** [comment]
I am only able to get the shapes for `values` and `base_values`:

values.shape: (4, )
base_values.shape: (4, 665)
feature_names.shape: AttributeError: 'list' object has no attribute 'shape'
data.shape: AttributeError: 'tuple' object has no attribute 'shape'

**Comment by remo-help** [comment]
Thanks. The feature_names are expected, the rest is curious.

`.values` are your actual shap values. The fact that they are of shape (4, ) is very odd. Did you only infer one row of data? Have you tried looking at your `shap_values.values` to see what's actually in the array?
`.data` suggests to me that the data you are passing to your model is passed as a tuple, is this correct?

**Comment by KuanWeiBeCool** [comment]
Thanks for checking! As I mentioned in the original post, I infer 4 instances, that is 4 rows of data. I think the reason why .values are of shape (4, ) is because every row has different tokens and thus different number of values, so numpy cannot use a single value to represent the second dimension. Here are the shapes if I select the individual rows:
![image](https://user-images.githubusercontent.com/64152297/183098663-9491e3f8-4582-40e1-999a-eaed46f021be.png)

Here is the actual values:
![image](https://user-images.githubusercontent.com/64152297/183098775-f384d151-f215-484c-8237-b042f5a464c9.png)

I am passing a string to the model. My model is a NLP classification model and it takes a string and output a classification. So it's not exactly tuple but similar. The `.data` object is a tuple of numpy array:

![image](https://user-images.githubusercontent.com/64152297/183099467-303a6b79-d4b4-42d5-8f90-9e5dd9f031f3.png)

**Comment by remo-help** [comment]
I see, and you can generate other plots with this Explanation object?

The error you're getting seems related to the `.output_names` parameter of the explanation object. Can you see if this parameter is defined by calling:
`shap_values.output_names`

**Comment by KuanWeiBeCool** [comment]
Yes, I can generate text plot, waterfall plot, bar plot without an issue.

`shap_values.output_names` is defined and contains a list of all the outputs:
![image](https://user-images.githubusercontent.com/64152297/183103077-f292ce9a-32f2-4b6f-927e-a78a23af231c.png)

**Comment by remo-help** [comment]
Hmm. Looking at the particular section of the code that fails here, it looks like it's trying to generate a new Explanation object along a passed index, by iterating over every item in the explanation object.
Since you pass it a string index, it executes the loop that starts on line 340 of `_explanation.py`. It grabs the dimension of the `output_names` parameter of the subslice that it is currently looking at (lines 443-447). It finds that `len(output_names_dims) ==1`. I'm not sure what that means in this context.
From there, it executes this strange line:
`t = np.argwhere(np.array(self.output_names) == t)[0][0]`
where `t` is the current indexed slice of the `Explanation` object that it is iterating over (as far as I can see).
Since it tries to to run a `np.argwhere` operation over `t` and grab the `[0][0]` indices of that array, I think the issue is simply that this operation is not built to deal with arrays that are not 2D. Your explanation object is a 3D array.
Basically, the only thing, _I believe_, you can do here, is actually modify the source code in this line to add an if-condition here that checks the dimensionality of the explanation object and if that dimensionality is 3 runs this line instead:
`t = np.argwhere(np.array(self.output_names) == t)[0][0][0]`

No guarantee that that will fix it, but it's a possibility.

If you want to do this, you can set up a fresh python environment. Fetch the repository, make your modifications. And then while in the repository, you can call: `pip install -e .` 
This will install the local version of the package as your SHAP version. Then you can try if that works. And if it does, heck, you can make a PR :)
If not, then I probably misread the error.

**Comment by KuanWeiBeCool** [comment]
I see. Guess I need to dive more into the source code then. Thanks for your help! 

By the way, do you know why in the documentation (https://shap.readthedocs.io/en/latest/example_notebooks/api_examples/plots/text.html) the slicing works? It's also supposed to be a 3D array since the explainer is run on multiple samples of text...

**Comment by remo-help** [comment]
The example shouldn't be a 3D array, since they slice it like a 2D array:

`shap.plots.bar(shap_values[:,"but"])`

Unless this is actually possible with a 3D explanation object, in that case you should be able to just call:

`shap.plots.bar(shap_values[:, 'mix'])`

**Comment by KuanWeiBeCool** [comment]
I see. I just have one last question if you don't mind. My intuition is that in the documentation example, since it's sentiment analysis, probably it selects just one label. I tried this method as well by selecting just one label: shap_values[:, :, 365], and it shrinks down to a 2D array (named `test`). However, when I try to slice this 2D array with `test[:, 'mix']`, it says `IndexError: too many indices for array`. Theoretically if it's an 2D array the code should work right, why it doesn't work in this case?

**Comment by KuanWeiBeCool** [comment]
Full error message:
![image](https://user-images.githubusercontent.com/64152297/183112228-e7b70137-7e6f-4bf1-b7e3-122d451b1b2c.png)

**Comment by remo-help** [comment]
`shap_values[:, :, 365]`
what you've done here is you've sliced out all the values for the feature at index 365.
So if you then call `test[:, 'mix']`, you're trying to slice it along a dimension that no longer exists.

Did you try `shap.plots.bar(shap_values[:, 'mix'])` jsut to be sure?

**Comment by KuanWeiBeCool** [comment]
Just to clarify, when I do shap_values[:, :, 365], I sliced out all the values for the **label** at index 365, not the feature. I have 665 labels here. It's a multi-label classification. `test` is still a 2D array with shape (4, None), where None is there because the number of tokens in each sample is different

Here is what happened when trying shap_values[:, 'mix], same error as above:
![image](https://user-images.githubusercontent.com/64152297/183115722-31f4f3a9-6f24-4b3f-9119-e05ca4f502ec.png)

**Comment by remo-help** [comment]
I see, this changes things quite a bit. The shap values I'm used to working with are sorted along these dimensions: `[rows:features]`, which is why I suggested the steps I suggested.

Where are the actual features in your object?

**Comment by KuanWeiBeCool** [comment]
The actual features are word tokens, and I converted them using TFIDF before feeding into my NLP model

**Comment by remo-help** [comment]
Yes, what I am asking is whether you know along which dimensions your features (i.e. the shap values for those features) are represented in your explanation object.

Since you are trying to slice the object using a feature ('mix') as an index, it's a relevant question along which dimension your features are represented

**Comment by KuanWeiBeCool** [comment]
Oh sorry I misunderstood your question. Let's say my `shap_values` has a dimension of (4, None, 665), the second dimension (which has None) represents the dimension of shap values for individual tokens. '4' represents the number of rows, 665 represents the number of labels. Does this answer your question?

**Comment by remo-help** [comment]
Sort of, I'm not sure I understand how a dimension of `None` can represent values for individual tokens.

Essentially, the problem is this: You are trying to re-arrange the object along an index representing a feature ('mix'), so you need to put that index in the right dimension (the dimension where that index can actually be found). It would seem that dimension is the 2nd dimension, but the shape of that dimension is `None`, which makes no sense to me

**Comment by KuanWeiBeCool** [comment]
The reason why the second dimension is `None`, I believe, is because each row has different number of tokens, and thus different number of shap values, as shown in the following image. Therefore, it is impossible to represent the second dimension using a single value. I also have an intuition that this is where the error comes from, but I am confused why it worked in the SHAP document, where it also runs on multiple number of rows of text data, and must also have different number of tokens for each row...
![image](https://user-images.githubusercontent.com/64152297/183131566-8250185d-5d31-439b-9c30-5ad0c84d5d0b.png)

**Comment by remo-help** [comment]
Hmm, If I were you I would run the example from the documentation and inspect that explanation object. It might give you some insight as to what's going on.

Are you using a custom predict function that takes in text and outputs logits like they do (and are you passing it to the explainer)? 
```
def f(x):
    tv = torch.tensor([tokenizer.encode(v, padding='max_length', max_length=500, truncation=True) for v in x]).cuda()
    outputs = model(tv)[0].detach().cpu().numpy()
    scores = (np.exp(outputs).T / np.exp(outputs).sum(-1)).T
    val = sp.special.logit(scores[:,1]) # use one vs rest logit units
    return val
```

I think the shape of your object is preventing shap from correctly slicing it along the index you're specifying. Since the documentation example is analogous task, you should run it and compare the objects it outputs vs yours.

**Comment by KuanWeiBeCool** [comment]
Yes I have a custom predict function similar to the one in the example. I'll try running it and see if I could get the right thing. Thanks a lot!

**Comment by remo-help** [comment]
I led you on a wrong path in the beginning, so it's the least I can do. Hope it works out

**Comment by KuanWeiBeCool** [comment]
Haha no worries, it is a confusing thing

**Comment by waikoreaweatherpjt** [comment]
when can you solve this error???

**Comment by adamhrin** [comment]
Hello @slundberg , currently this error is still a problem and even the example itself in the notebook does not work with the latest version of the package. The error happens in slicer's [`__getitem__`](https://github.com/interpretml/slicer/blob/1e6fe511cd888a6cf092087533fc003e2c028c7e/slicer/slicer.py#L112) when `tracked` represents `"values"`, because the `tracked.max_dim` parameter is smaller than `len(slicer_index)` and that subsequently results in the error. I believe this is due to how _explanation's [`__getitem__`](https://github.com/shap/shap/blob/1da1fb7d526076394d108ffea593f16c0565d356/shap/_explanation.py#L389) handles creating the `new_values` and `new_data` objects. The resulting data and values (in the case of the [notebook example](https://github.com/shap/shap/blob/master/notebooks/api_examples/plots/text.ipynb)) are just 1d arrays `['but']` and `[-0.0713973339875198]` respectively, and these dimensions are smaller than original values and data had. I don't know how to go about fixing this problem though. I would appreciate your help with this! It would be great to have such feature for further analysis of texts by presence of tokens.
